CombinatorialGames/Game/Loopy/Basic.lean

```diff
@@ -612,4 +649,432 @@ theorem off_sub_off : off - off = dud := by
+def leftMoves (x : MulTy α β) : Set (MulTy α β) :=
+  ⋃ y ∈ x, (x.erase y + ·) '' leftMovesSingle leftMovesα rightMovesα leftMovesβ rightMovesβ y
+
+/-- The set of right moves of `Σ ±xᵢ * yᵢ` are `zᵢ + Σ ±xⱼ * yⱼ` for all `i`, where `cᵢ` is a right
+move of `±xᵢ * yᵢ`, and the summation is taken over indices `j ≠ i`. -/
+def rightMoves (x : MulTy α β) : Set (MulTy α β) :=
+  ⋃ y ∈ x, (x.erase y + ·) '' rightMovesSingle leftMovesα rightMovesα leftMovesβ rightMovesβ y
+
+-- TODO: upstream?
+theorem _root_.Multiset.iUnion_map {α β γ} (m : Multiset α) (f : α → β) (g : β → Set γ) :
+    ⋃ x ∈ m.map f, g x = ⋃ x ∈ m, g (f x) := by
+  simp
+
+theorem leftMoves_neg (x : MulTy α β) :
+    leftMoves leftMovesα rightMovesα leftMovesβ rightMovesβ (-x) =
+    -rightMoves leftMovesα rightMovesα leftMovesβ rightMovesβ x := by
+  rw [leftMoves, rightMoves, neg_def, Multiset.iUnion_map]
+  simp [← image_neg_eq_neg, image_iUnion, image_image, leftMovesSingle_neg, ← neg_def]
+
+theorem rightMoves_neg (x : MulTy α β) :
+    rightMoves leftMovesα rightMovesα leftMovesβ rightMovesβ (-x) =
+    -leftMoves leftMovesα rightMovesα leftMovesβ rightMovesβ x := by
+  rw [leftMoves, rightMoves, neg_def, Multiset.iUnion_map]
+  simp [← image_neg_eq_neg, image_iUnion, image_image, rightMovesSingle_neg, ← neg_def]
+
+variable {α₁ β₁ α₂ β₂ : Type*}
+  {leftMovesα₁ rightMovesα₁ : α₁ → Set α₁} {leftMovesβ₁ rightMovesβ₁ : β₁ → Set β₁}
+  {leftMovesα₂ rightMovesα₂ : α₂ → Set α₂} {leftMovesβ₂ rightMovesβ₂ : β₂ → Set β₂}
+  (f : α₁ → α₂) (g : β₁ → β₂)
+
+/-- Map `MulTy α₁ β₁` to `MulTy α₂ β₂` using `f : α₁ → α₂` and `g : β₁ → β₂` in the natural way. -/
+def map : MulTy α₁ β₁ → MulTy α₂ β₂ :=
+  Multiset.map (Prod.map id (Prod.map f g))
+
+@[simp]
+theorem map_add (x y) : map f g (x + y) = map f g x + map f g y :=
+  Multiset.map_add ..
+
+theorem map_erase [DecidableEq α₁] [DecidableEq β₁] [DecidableEq α₂] [DecidableEq β₂]
+    {x : MulTy α₁ β₁} {y} (hy : y ∈ x) :
+    map f g (x.erase y) = (map f g x).erase (y.1, f y.2.1, g y.2.2) :=
+  Multiset.map_erase_of_mem _ _ hy
+
+@[simp]
+theorem map_mulOption (b x y) :
+    map f g (mulOption b x y) = mulOption b (Prod.map f g x) (Prod.map f g y) := by
+  simp [mulOption, map, Prod.map]
+
+variable {f g}
+
+set_option maxHeartbeats 1000000 in
+theorem leftMovesSingle_comp_prodMap
+    (hlf : leftMovesα₂ ∘ f = Set.image f ∘ leftMovesα₁)
+    (hrf : rightMovesα₂ ∘ f = Set.image f ∘ rightMovesα₁)
+    (hlg : leftMovesβ₂ ∘ g = Set.image g ∘ leftMovesβ₁)
+    (hrg : rightMovesβ₂ ∘ g = Set.image g ∘ rightMovesβ₁) :
+    leftMovesSingle leftMovesα₂ rightMovesα₂ leftMovesβ₂ rightMovesβ₂ ∘ Prod.map id (Prod.map f g) =
+    image (map f g) ∘ leftMovesSingle leftMovesα₁ rightMovesα₁ leftMovesβ₁ rightMovesβ₁ := by
+  simp_rw [funext_iff, Function.comp_apply, leftMovesSingle, leftMovesSet] at *
+  rintro ⟨(_ | _), x⟩ <;> aesop
+
+-- TODO: find a faster proof using the previous theorem?
```

Comment:
(thats if we prove that `rightMovesSingle` is an additive monoid homomorphism beforehand, which is a bit more annoying than this proof)

Reply:
My idea was to convert this into the `leftMoves` version by using negation, but that didn't quite work out.

There might be some other simple proof that avoids `aesop`, so I'll leave the TODO comment here.